File: 3744-minimum-operations-to-make-array-elements-zero/minimum-operations-to-make-array-elements-zero.java
class Solution {

    private long get(int num) {
        if (num <= 0) return 0;
        long cnt = 0;
        int i = 0;
        long base = 1;
        while (base <= num) {
            long end = Math.min(base * 4 - 1, (long) num);
            long count = end - base + 1;
            cnt += (long) (i + 1) * count;  // each number in this block takes (i+1) steps
            i++;
            base *= 4;
        }
        return cnt;
    }

    public long minOperations(int[][] queries) {
        long res = 0;
        for (int[] q : queries) {
            long count1 = get(q[1]);
            long count2 = get(q[0] - 1);
            long totalSteps = count1 - count2;
            res += (totalSteps + 1) / 2; // ceil division
        }
        return res;
    }
}

File: 3744-minimum-operations-to-make-array-elements-zero/minimum-operations-to-make-array-elements-zero.py
from typing import List
import math

class Solution:
    def minOperations(self, queries: List[List[int]]) -> int:
        # helper: prefix sum of steps up to n
        def prefix_steps(n: int) -> int:
            if n == 0:
                return 0
            total = 0
            i = 0
            while True:
                left = 4**i
                right = 4**(i+1) - 1
                if left > n:
                    break
                start = left
                end = min(n, right)
                count = end - start + 1
                total += count * (i + 1)  # each in this bucket has (i+1) steps
                i += 1
            return total

        ans = 0
        for l, r in queries:
            total_steps = prefix_steps(r) - prefix_steps(l - 1)
            ans += (total_steps + 1) // 2   # ceil division
        return ans
